concepts/pointers/introduction.md

```diff
@@ -1,3 +1,149 @@
+
+## Variables and Memory
+
+Let's say we have a regular integer variable `a`:
+
+```go
+var a int
+```
+
+When we declare a variable, Go has to find a place in memory to place its value. Conviniently, when we need to fetch the value stored in that piece of memory, we can just refer to it by the variable name.
+
+For instance, when we do:
+
+```go
+a + 2
+```
+
+We are effectively fetching the value stored in the memory associated with the variable `a` and adding 2 to it.
+
+Similarly, when we need to change the value in the piece of memory of `a`, we can use the variable name to do an assignment:
+
+```go
+a = 3
+```
+
+The piece of memory that `a` is associated with now will have the value `3`.
+
+## Pointers
+
+While variables allow us to refer to values in memory, sometimes it's useful to know the memory address of the value that the variable refers. Pointers hold the memory addresses of those values. A pointer declaration looks like this:
+
+```go
+var p *int // 'p' can hold the memory address of an integer
+```
+
+Here we declaring a variable `p` of type `*int`. The type `*int` means "pointer to int". In other words, we are saying that `p` is a pointer to int, which means that `p` will hold the memory address of an integer. The zero value of pointers is `nil`. A `nil` pointer holds no memory address.
+
+### Getting a pointer to a variable
+
+The operator `&` gives the memory address of its operand. This means that to get the memory address of the value of a variable `a`, we can do `&a`. `&a` will create a pointer to the value of `a`:
+
+```go
+var a int
+a = 2
+
+var p *int
+p = &a // the variable 'p' wil hold the memory address of 'a'
+```
+
+### Accessing the value via a pointer (dereferencing)
+
+When we have a pointer, we might want to know the value that is in the memory address of the pointer. We can do this using the `*` operator:
+
+```go
+var a int 
+a = 2
+
+var p *int 
+p = &a // the variable 'p' wil hold the memory address of 'a'
+
+var b int
+b = *p // b == 2
+```
+
+The operation `*p` fetches the value stored at the memory address stored in `p`. This operation is often called "dereferencing".
+
+We can also use the derefering operator to assign a new value to the memory address referenced by the pointer:
+
+```go
+var a int
+a = 2   // declare int variable 'a' and assign it the value of 2
+
+var pa *int
+pa = &a          // 'pa' now refers to the memory address of 'a'  
+*pa = *pa + 2    // increment by 2 the value at memory address 'pa'
+
+fmt.Println(a)   // Output: 4
+                 // 'a' will have the new value that was changed via the pointer!
+```
+
+Assigning to `*pa` will change the value stored at the memory address `pa` holds. Since `pa` holds the memory address of `a`, by assigning to `*pa` we are effectively changing the value of `a`! 
+
+A note of caution however: always check if a pointer is not `nil` before dereferencing. Dereferencing a `nil` pointer will make the program panic at runtime!
```

Comment:
```suggestion
A note of caution however: always check if a pointer is not `nil` before dereferencing. Dereferencing a `nil` pointer will make the program panic at runtime!
```

Could we use a different word to "panic" here? Maybe "crash". I'm not sure someone coming from a different language (e.g. Ruby) would be familiar with the word, and so I have images in my mind of someone running around panicking, rather than the more technical usage :)

Reply:
In Go, a crash (in the sense of unhandled exception) is called a panic, that was the main reason for the use of panic there.

But I it doesn't hurt to make the introduction more inclusive and use a more general term.

Thanks for the suggestion.

Reply:
Fixed.